include/tvm/runtime/threading_backend.h

```diff
@@ -0,0 +1,50 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file threading_backend.h
+ * \brief Utilities for manipulating thread pool threads.
+ */
+#ifndef TVM_RUNTIME_THREADING_BACKEND_H_
+#define TVM_RUNTIME_THREADING_BACKEND_H_
+
+#include <functional>
+#include <memory>
+#include <vector>
+
+namespace tvm {
+namespace runtime {
+namespace threading {
+
+class ThreadGroup {
+ public:
+  class Impl;
+
+   /*!
+    * \brief Starts and manages threads which back a thread pool.
+    *
+    * \param num_workers The number of worker threads to start.
+    * \param worker_callback A callback which is run in its own thread.
+             Receives the worker_id as an argument.
+    * \param include_main_thread Whether to use the main thread as a worker.
+    *        If this is set to true, one worker will re-use the main thread.
+    */
+  ThreadGroup(int num_workers, std::function<void(int)> worker_callback,
```

Comment:
remove constructor, instead add a Create function with same signature. By default c++ constructor cannot throw exceptions, which means error during constructor will cause seg directly

Reply:
the benefit of stack allocable is neglible

Reply:
so you're definitely against having a constructor even though it has the same exception handling mechanisms as a factory function?

Reply:
OK, after checking http://en.cppreference.com/w/cpp/language/noexcept_spec seems standard says that non-default constructors can throw, so I am fine with the constructor if that is the case